### DISCUSSION TITLE
Task durations

### DISCUSSION THREAD

**Discussion by vakker** [shop-planner/shop3#109] [Q&A] [answered]
Hi,

Does SHOP3 support task durations? 
The documentation doesn't mention it:
```lisp
(:op head
       [:precond precondition]
       [:delete delete-list]
       [:add add-list]
       [:cost cost-fn])
```

But in the PDDL examples (e.g. `IPC-2002/Depots/SimpleTime/stdepots.pddl`) there are actions with `:duration (= ?duration 10)`. 

PDDL should be strictly less expressive than SHOP3, so I assume there's a way to capture that without PDDL, right?

Thanks

**Comment by rpgoldman** [comment]
Hi, @vakker -- I'm changing this from an Issue to a Discussion, because I'd like Q&A's to be searchable, but Issues might be harder to find.  More to follow.

**Reply by vakker** [reply]
Makes sense, sorry about that! Thanks for looking into it!

**Reply by rpgoldman** [reply]
Answering the question now....

**Comment by rpgoldman** [comment, marked as answer]
SHOP3 does not have temporal planning fully built in.  There are a couple of ways to do temporal (sometimes called "durative") planning with SHOP3:

1. There's a method for dealing with it that is described in an early paper on SHOP2 that the folks at University of Maryland figured out.  See Section 4.2 in [this paper](https://www.researchgate.net/publication/220543221_SHOP2_An_HTN_planning_system)
2. I developed a different technique that is described in [this paper](https://rpgoldman.goldman-tribe.org/papers/icaps06-short.pdf). If you are interested in this method, I can dig up the `temporal-shop2` system and put it into SHOP3. It will likely need some maintenance to work with SHOP3, but it shouldn't take much.

**Reply by vakker** [reply]
Thanks, that helps. If you could provide the example `temporal-shop2` that would be highly appreciated!

**Reply by rpgoldman** [reply]
I'm getting it packaged up now. I got bogged down with fixing the Blocks domain, which in turn led me to discover that the GitHub actions were all broken! 🤯

**Comment by rpgoldman** [comment]
I have published the temporal-shop code to this project: https://github.com/shop-planner/temporal-shop

Note that this code *has not been updated to SHOP3* so almost certainly will not work out of the box.  I aspire to getting it updated, but don't have any time to dig into it right now. Any updates would be very welcome as pull requests!

**Reply by vakker** [reply]
Thanks! Is that repo public? That link gives 404 for me.

**Reply by rpgoldman** [reply]
Sorry! It was meant to be public; I didn't realize it would default to private. Fixed now.
